### PR TITLE
Fixes startup exceptions happens which happens before dependent services is up 

### PR DESCRIPTION
**Pull request by kotharironak**
Address the issue - https://github.com/hypertrace/query-service/issues/29

### CODE REVIEW

File: hypertrace-service/src/main/java/org/hypertrace/service/HypertraceUIServerTimerTask.java [hunk 1]
```diff
@@ -53,30 +54,36 @@
...
   private int pinotServerPort;
   private int pinotControllerPort;
+  private boolean isPinotUp;
 
 
-  public HypertraceUIServerTimerTask(Config appConfig, HypertraceUIServer uiServer, String defaultTenant) {
-    maxRetries = appConfig.hasPath(RETRIES_CONFIG) ? appConfig.getInt(RETRIES_CONFIG) : DEFAULT_RETRIES;
-    timeout = appConfig.hasPath(TIMEOUT_CONFIG) ? appConfig.getInt(TIMEOUT_CONFIG) : DEFAULT_TIMEOUT;
+  public HypertraceUIServerTimerTask(Config appConfig, HypertraceUIServer uiServer,
+      String defaultTenant) {
```
Comment (kotharironak) [inline hypertrace-service/src/main/java/org/hypertrace/service/HypertraceUIServerTimerTask.java:61]: Formatting changed, applied google style

File: hypertrace-service/src/main/java/org/hypertrace/service/HypertraceUIServerTimerTask.java [hunk 2]
```diff
@@ -53,30 +54,36 @@
...
+    maxRetries =
+        appConfig.hasPath(RETRIES_CONFIG) ? appConfig.getInt(RETRIES_CONFIG) : DEFAULT_RETRIES;
+    timeout =
+        appConfig.hasPath(TIMEOUT_CONFIG) ? appConfig.getInt(TIMEOUT_CONFIG) : DEFAULT_TIMEOUT;
     interval = appConfig.hasPath(INTERVAL) ? appConfig.getInt(INTERVAL) : DEFAULT_INTERVAL;
-    startPeriod = appConfig.hasPath(START_PERIOD) ? appConfig.getInt(START_PERIOD) : DEFAULT_START_PERIOD;
+    startPeriod =
```
Comment (kotharironak) [inline hypertrace-service/src/main/java/org/hypertrace/service/HypertraceUIServerTimerTask.java:67]: Formatting changed, applied google style

File: hypertrace-service/src/main/java/org/hypertrace/service/HypertraceUIServerTimerTask.java [hunk 3]
```diff
@@ -53,30 +54,36 @@
...
+        appConfig.hasPath(START_PERIOD) ? appConfig.getInt(START_PERIOD) : DEFAULT_START_PERIOD;
 
     pinotSeverHost = appConfig.hasPath(PINOT_SERVER_HOST) ?
-            appConfig.getString(PINOT_SERVER_HOST) : DEFAULT_PINOT_SERVER_HOST;
+        appConfig.getString(PINOT_SERVER_HOST) : DEFAULT_PINOT_SERVER_HOST;
```
Comment (kotharironak) [inline hypertrace-service/src/main/java/org/hypertrace/service/HypertraceUIServerTimerTask.java:71]: Formatting changed, applied google style

File: hypertrace-service/src/main/java/org/hypertrace/service/HypertraceUIServerTimerTask.java [hunk 4]
```diff
@@ -53,30 +54,36 @@
...
     this.numRetries = 0;
     this.defaultTenant = defaultTenant;
     this.startTime = Instant.now().toEpochMilli();
+    this.isPinotUp = false;
 
     client = GatewayServiceGrpc.newBlockingStub(ManagedChannelBuilder.forAddress(
-            "localhost", appConfig.getInt("service.port")).usePlaintext().build())
-            .withCallCredentials(RequestContextClientCallCredsProviderFactory
-                    .getClientCallCredsProvider().get());
+        "localhost", appConfig.getInt("service.port")).usePlaintext().build())
```
Comment (kotharironak) [inline hypertrace-service/src/main/java/org/hypertrace/service/HypertraceUIServerTimerTask.java:84]: Formatting changed, applied google style

File: hypertrace-service/src/main/java/org/hypertrace/service/HypertraceUIServerTimerTask.java [hunk 5]
```diff
@@ -92,47 +99,70 @@ public void run() {
...
+                "It seems dependent data service [pinot] is not yet up. " +
+                "will retry after [%s] seconds", numRetries, interval));
+        return;
+      }
+
+      if (executeHealthCheck()) {
         cancel();
         LOGGER.info(String.format("Stack is up after [%s] attempts, and duration [%s] in millis.",
-                numRetries, Instant.now().toEpochMilli() - startTime));
+            numRetries, Instant.now().toEpochMilli() - startTime));
```
Comment (kotharironak) [inline hypertrace-service/src/main/java/org/hypertrace/service/HypertraceUIServerTimerTask.java:134]: Formatting changed, applied google style

File: hypertrace-service/src/main/java/org/hypertrace/service/HypertraceUIServerTimerTask.java [hunk 6]
```diff
@@ -92,47 +99,70 @@ public void run() {
...
     } finally {
       numRetries++;
     }
   }
 
   private boolean executeHealthCheck() {
     SpansResponse response = GrpcClientRequestContextUtil.executeInTenantContext(defaultTenant,
-            () -> client.withDeadline(Deadline.after(timeout, TimeUnit.SECONDS))
-                    .getSpans(buildSpanRequest()));
+        () -> client.withDeadline(Deadline.after(timeout, TimeUnit.SECONDS))
```
Comment (kotharironak) [inline hypertrace-service/src/main/java/org/hypertrace/service/HypertraceUIServerTimerTask.java:154]: Formatting changed, applied google style

File: hypertrace-service/src/main/java/org/hypertrace/service/HypertraceUIServerTimerTask.java [hunk 7]
```diff
@@ -92,47 +99,70 @@ public void run() {
...
   }
 
   private SpansRequest buildSpanRequest() {
     return SpansRequest.newBuilder()
-            .setStartTimeMillis(System.currentTimeMillis() - TimeUnit.SECONDS.toMillis(10))
-            .setEndTimeMillis(System.currentTimeMillis())
-            .addSelection(QueryExpressionUtil.getColumnExpression("EVENT.id"))
-            .setLimit(1)
-            .build();
+        .setStartTimeMillis(System.currentTimeMillis() - TimeUnit.SECONDS.toMillis(10))
```
Comment (kotharironak) [inline hypertrace-service/src/main/java/org/hypertrace/service/HypertraceUIServerTimerTask.java:161]: Formatting changed, applied google style

File: hypertrace-service/src/main/java/org/hypertrace/service/HypertraceUIServerTimerTask.java [hunk 8]
```diff
@@ -157,7 +187,8 @@ private boolean executePinotHealthCheck() throws Exception {
   private boolean executeBrokerRegistrationCheck() throws Exception {
     HttpURLConnection con = null;
     try {
-      URL url = new URL(String.format("http://%s:%s/brokers/tenants", pinotSeverHost, pinotControllerPort));
+      URL url = new URL(
```
Comment (kotharironak) [inline hypertrace-service/src/main/java/org/hypertrace/service/HypertraceUIServerTimerTask.java:190]: Formatting changed, applied google style

**Comment by jcchavezs** [comment]
I think keeping a style consistent is good. Lets please add a check in a
next PR all with the fix for all files.

File: hypertrace-service/src/main/java/org/hypertrace/service/HypertraceUIServerTimerTask.java [hunk 9]
```diff
@@ -92,47 +99,70 @@ public void run() {
     try {
       if (numRetries >= maxRetries) {
         cancel();
-        LOGGER.info(String.format("Max out attempts [%s] in checking bootstrapping status. Manually check " +
+        LOGGER.info(String
+            .format("Max out attempts [%s] in checking bootstrapping status. Manually check " +
```
Comment (satish-mittal) [inline hypertrace-service/src/main/java/org/hypertrace/service/HypertraceUIServerTimerTask.java:103]:
```suggestion
            .format("Exhausted maximum attempts [%s] in checking bootstrapping status. Manually check " +
```

File: hypertrace-service/src/main/java/org/hypertrace/service/HypertraceUIServerTimerTask.java [hunk 10]
```diff
@@ -92,47 +99,70 @@ public void run() {
...
                 "the status of data service [pinot].", numRetries));
         uiServer.start();
         return;
       }
 
-      if (executePinotHealthCheck() && executeBrokerRegistrationCheck() && executeHealthCheck()) {
+      /**
+       * We will check for e2e spans request in the next time cycle after the pinot is up. We are
+       * using a zookeeper based pinot client in query service which needs to brokers to be registered.
```
Comment (satish-mittal) [inline hypertrace-service/src/main/java/org/hypertrace/service/HypertraceUIServerTimerTask.java:111]:
```suggestion
       * using a zookeeper based pinot client in query service which needs brokers to be registered.
```

File: hypertrace-service/src/main/java/org/hypertrace/service/HypertraceUIServerTimerTask.java [hunk 11]
```diff
@@ -92,47 +99,70 @@ public void run() {
...
+       * So, instead of checking immediately after the pinot is up, we are adding buffer time for
+       * it to be fully functional. This avoids any confusion between any pre-configured errors
+       * messages during start time, see an issue - https://github.com/hypertrace/query-service/issues/29
+       * */
+      if (!isPinotUp && executePinotHealthCheck() && executeBrokerRegistrationCheck()) {
+        isPinotUp = true;
+        LOGGER
+            .warn(String.format("Finished an attempt [%s] in checking for bootstrapping status. " +
+                "It seems dependent data service [pinot] is up now. " +
+                "will check for span status after [%s] seconds", numRetries, interval));
```
Comment (satish-mittal) [inline hypertrace-service/src/main/java/org/hypertrace/service/HypertraceUIServerTimerTask.java:121]:
```suggestion
                "Will check for span status after [%s] seconds", numRetries, interval));
```

File: hypertrace-service/src/main/resources/configs/hypertrace-service/application.conf [hunk 12]
```diff
@@ -6,7 +6,7 @@ service.admin.port = 9002
 hypertraceUI {
   port = 2020
   init.waittime {
-    timeout = 2
+    timeout = 5
```
Comment (kotharironak) [inline hypertrace-service/src/main/resources/configs/hypertrace-service/application.conf:9]:
```suggestion
    timeout = 10
```

**Review by laxmanchekka — APPROVED** [review]

**Comment by kotharironak** [comment]
It helps in fixing the issue - https://github.com/hypertrace/hypertrace-service/issues/40

**Pull request merged**